### PR TITLE
add retrying client

### PR DESCRIPTION
**Pull request by MarioBassem**
https://github.com/threefoldtech/terraform-provider-grid/issues/277

### CODE REVIEW

**Review by muhamadazmy — CHANGES_REQUESTED** [review]
## Review
I have checked the issue and I don't believe "retrying" is the right fix for this issue for the following reason:
- Retrying (I suppose) will not reconnect. Normally when a connection is closed, it's closed. A new connection is needed
- I think the main problem (why you have this issue) is that the connection is used in a wrong way. The right way to use this library is by using the connection manager 

The `Manager` (available in this package) is a naive connection pool (which doesn't actually pool the connections but should be) is a way to create and destroy connections on demand. Basically like
```go
con := manager.Substrate();
defer con.Close();

// use the connection for few operations that are not long lived.
```

This way, u know you always have a valid connection. On connection close the Manager can decide either to actually close the connection or keep it for reuse (I know it's currently just close it but again this can/should be fixed)

A bad usage of this Manager is to get one connection and just keep it alive for the entire life of the application. That is wrong because once the connection is dead there is noway to bring it back (i assume also retrying won't fix it unless it's explicitly implemented in the underlying client to reconnect).

I would honestly prefer to implement instead to:
- Implement decent pooling in the `Manager`
- Fix the "user" of this library to create connections on demand and not use the same connection for the entire app.

@OmarElawady what do you think?

**Comment by OmarElawady** [comment]
When the connection is kept idle for too long (~3 mins), the connection gets closed and any query fails. After failure the client's connection is marked internally as broken, and it reconnects before performing any calls. That's why retrying one time makes the failure hidden and not returned to the library's user. So this fix looks like it's convenient in any case (same as https://github.com/threefoldtech/substrate-client/blob/master/utils.go#L132?)

Implementing the pool and making the user hold a connection for a short time would result in the same behavior, right? since pools shouldn't close the connection after it's returned to them. And the pool won't affect terraform (and zos?), since its usage is sequential?

And I have a question regarding how pool is used. There are two ways I saw to use pools with clients:
1. To make a pool object with a method to get a client connection, and then this connection is used and returned to the pool once it's no longer needed. Like in https://github.com/gomodule/redigo
2.  To make the pool embedded in the client, so the client methods is used directly and it gets a connection from the pool internally and uses it. Like in github.com/go-redis/redis

Honestly, I don't see what is the benefit of doing it the first way, it makes the usage more difficult in the sense that the user must code carefully to not hold the connection for too long. The second way makes the connections held for the minimal amount possible and it's totally transparent to the user.

What to do in terraform, depends on which way the pool is going to be implemented. The first way requires a decent amount of refactoring while the second way will make the usage almost as is.

So TLDR; IMHO, I think the retrying is useful in any case and it'll solve terraform's problem. For the pool, would be great to have your input on which way it should be implemented and depending on that terraform would be adjusted.

**Comment by muhamadazmy** [comment]
Answering your questions here:
- Pools must make sure the connection is valid before they actually hand the connection to the caller. 
  - Pools are usually configured with timeout for idle connection. Hence you can configure if a connection is open "idle" for this amount of seconds, the connection is considered dead, closed and cleaned out.
  - For connection that didn't hit this timeout yet, they are usually tested before handed out the test is usually a "fast" test so it depends on the protocol what u do.

As you said, the client itself can hide the Pool internally hence each call to a client method just get a connection from the internal pool, and then put it back hence the pool implementation can decide when to create a new connection and when to reuse.

That being said, I don't mind your solution at all. My only concern was if "retrying" would actually reconnect because that is very implementation specific. 

If this is tested enough, merge it!

**Review by muhamadazmy — APPROVED** [review]
